src/analyses/basePriv.ml

```diff
@@ -1208,6 +1307,23 @@ struct
       ) st.cpa st
 
   let threadenter = startstate_threadenter startstate
+
+  let invariant_global getg g =
+    match g with
+    | `Left _ -> (* mutex *)
+      Invariant.none
+    | `Right g' -> (* global *)
+      ValueDomain.invariant_global (fun x ->
+          let weaks = G.weak (getg (V.global x)) in
+          let d_weak = GWeak.fold (fun s' gweakw' acc ->
+              GWeakW.fold (fun w' v acc ->
+                  VD.join v acc
+                ) gweakw' acc
+            ) weaks (VD.bot ())
+          in
+          let d_init = GWeakW.find lockset_init (GWeak.find (Lockset.empty ()) weaks) in
+          VD.join d_weak d_init
+        ) g'
```

Comment:
Managed to extract bunch of the latter ones into one shared implementation. During that I realized that these separate `d_init` values are unnecessary because the fold over all weaks also includes the init lockset ones. I had just copied those parts from `read_global` and removed some filtering.

Also implemented it for #789 now.